### ISSUE TITLE
Clarification about CT values / density

### ISSUE THREAD

**Issue by hemmer** [eigenvivek/DiffDRR#100]
Hey super interesting project, thanks for making the source available! I've been running the examples for the example CT and was wondering if you could clarify a couple of points / check if you agree with my logic.

1. pydicom doesn't by default apply rescale / intercept to the raw pixel data, so instead of the CT values `ds.pixel_array` contains raw values (i've plotted this below). This has the effect for the example data in the repo of keeping values positive which is presumably a requirement for sensible DRR generation. PyDICOM advises _against_ using raw values though, see [here](https://pydicom.github.io/pydicom/stable/old/working_with_pixel_data.html).

**Histogram of raw values (padding not corrected just for demonstration)**
![image](https://user-images.githubusercontent.com/915048/212289943-532ef0c1-ce35-4f76-a994-af42206f397d.png)

2. I believe you _could_ load DICOM for which rescale/intercept are 1/0 i.e. the raw values are signed and therefore can be negative. I ran into this loading some other data, but you could simulate this scenario by [applying the correction](https://pydicom.github.io/pydicom/stable/old/working_with_pixel_data.html#modality-lut-or-rescale-operation) to get CT values as `volume[:, :, idx] = apply_modality_lut(ds.pixel_array, ds)`. I've done this below, and as expected having negative values produces a different DRR.

**Histogram of CT values (padding not corrected just for demonstration)**
![image](https://user-images.githubusercontent.com/915048/212291136-07452c5f-b250-47a9-9fc8-eb00122af7d1.png)
**Resulting DRR not visually correct**
![image](https://user-images.githubusercontent.com/915048/212291165-fbd64bc4-bfb3-4a1b-b6ac-7e9038d12e02.png)

3. If the above correct, is it sensible to a) use CT rather than raw values and b) then apply a consistent rescaling of to keep these CT values positive? In fact I think converting from CT to density is what is done in some contexts, e.g. https://github.com/arcadelab/deepdrr/blob/main/deepdrr/vol/volume.py#L211 (no affiliation with that project).

**Comment by eigenvivek** [comment]
Hi! Yup, I ran into this issue early in development too.

The "box shadow artifact" you see in the produced DRR is not correct. It's a result of voxels of pure air being stored as negative numbers (looks like -2^10 or -2^11 in your histograms). Since a DRR is basically a line integral through the volume, every intersection of a ray with air causes a subtraction. Ultimately, this results in a box shape being subtracted out of the output DRR.

My solution was to take all negative air values and map them to zero:
https://github.com/eigenvivek/DiffDRR/blob/f375695a7cf0bc5633b8a011d2f65c33586a9d7d/diffdrr/data/load.py#L41-L42

**Other corrections:**
- Never heard of `modality_lut` before, but the fact that it preserves negative values makes me not like it for this application. Could definitely just convert air values to zero after I've done this.
- The density conversion seems interesting too, although I don't know why it's piecewise linear.

It could be useful to include all these corrections as options in the `load` function. 

Does this help answer your question?

**Comment by hemmer** [comment]
Thanks for the response, very useful. I certainly can see the effect of negative values on the DRR. However I don't think it's correct (in general) to stick to the raw values. They are all positive in the example DICOM, but not in general. From [here](https://pydicom.github.io/pydicom/stable/old/working_with_pixel_data.html), you can see the type can be signed or unsigned and I have come across many DICOM files that I've loaded where the raw values can be negative. So in summary using raw values doesn't guarantee positivity (needed for DRR generation). 

However I concede you don't want to use the actual CT values though, and that you want air to not contribute to the line integral (i.e. be zero). So conversion to CT values can only be an intermediate step, further normalisation that makes air 0 and remaining values positive is needed. I'm not best to advise on that, but I have a suggestion below that is at least should reproduce your current results.

I did see the `correct zero`, however I don't agree that's it's completely doing as you say. The CT scan has padding (the outer circular halo) that can have large negative values (e.g. -2048). The `correct_zero` code currently reassigns _padding_ regions to 0, which works specifically for the example DICOM but not in general. I agree that care is needed to avoid padding affecting the DRR. A tool like Slicer3D is useful to visualise this:

![image](https://user-images.githubusercontent.com/915048/212344270-55fda274-165c-4a64-8284-681e8cc38c56.png)


![image](https://user-images.githubusercontent.com/915048/212344897-dd9d8b89-9b46-4ad5-9a04-2589b73ceb47.png)


So in summary I think my advice would be:

* standardize CT loading using modality_lut (to avoid the code only working on a subset of CT DICOM, i.e. signed only)
* _then_ apply a HU value -> density correction (I think something like `1024 + max(-1024, vol)` would reproduce your current settings), but other mappings are possible

Does this make sense?
